icu4j/main/classes/core/src/com/ibm/icu/impl/number/LongNameHandler.java

```diff
@@ -61,14 +67,241 @@ static String getWithPlural(String[] strings, StandardPlural plural) {
+     *   joinerChar contains the space character (if any) that separated the
+     *   placeholder from the rest of the pattern. Otherwise, joinerChar is set
+     *   to NUL. Only one space character is considered.
+     */
+    private static ExtractCorePatternResult extractCorePattern(String pattern) {
+        ExtractCorePatternResult result = new ExtractCorePatternResult();
+        result.joinerChar = 0;
+        int len = pattern.length();
+        if (pattern.startsWith("{0}")) {
+            result.placeholderPosition = PlaceholderPosition.BEGINNING;
+            if (len > 3 && Character.isSpaceChar(pattern.charAt(3))) {
+                result.joinerChar = pattern.charAt(3);
+                result.coreUnit = pattern.substring(4);
+            } else {
+                result.coreUnit = pattern.substring(3);
+            }
+        } else if (pattern.endsWith("{0}")) {
+            result.placeholderPosition = PlaceholderPosition.END;
+            if (Character.isSpaceChar(pattern.charAt(len - 4))) {
+                result.coreUnit = pattern.substring(0, len - 4);
+                result.joinerChar = pattern.charAt(len - 4);
+            } else {
+                result.coreUnit = pattern.substring(0, len - 3);
+            }
+        } else if (pattern.indexOf("{0}", 1) == -1) {
+            result.placeholderPosition = PlaceholderPosition.NONE;
+            result.coreUnit = pattern;
+        } else {
+            result.placeholderPosition = PlaceholderPosition.MIDDLE;
+            result.coreUnit = pattern;
+        }
+        return result;
+    }
+
     //////////////////////////
     /// BEGIN DATA LOADING ///
     //////////////////////////
 
+    // Gets the gender of a built-in unit: unit must be a built-in. Returns an empty
+    // string both in case of unknown gender and in case of unknown unit.
+    private static String getGenderForBuiltin(ULocale locale, MeasureUnit builtinUnit) {
+        ICUResourceBundle unitsBundle = (ICUResourceBundle) UResourceBundle.getBundleInstance(ICUData.ICU_UNIT_BASE_NAME, locale);
+
+        StringBuilder key = new StringBuilder();
+        key.append("units/");
+        key.append(builtinUnit.getType());
+        key.append("/");
+
+        // Map duration-year-person, duration-week-person, etc. to duration-year, duration-week, ...
+        // TODO(ICU-20400): Get duration-*-person data properly with aliases.
+        if (builtinUnit.getSubtype() != null && builtinUnit.getSubtype().endsWith("-person")) {
+            key.append(builtinUnit.getSubtype(), 0, builtinUnit.getSubtype().length() - 7);
+        } else {
+            key.append(builtinUnit.getSubtype());
+        }
+        key.append("/gender");
+
+        try {
+            ICUResourceBundle stackBundle =
+                (ICUResourceBundle)unitsBundle.getWithFallback(key.toString());
+            return stackBundle.getString();
+        } catch (Exception e) {
```

Comment:
Optional: it is more idiomatic in Java to catch specific types of exceptions

Reply:
Discussed, and briefly surveyed other "catch (Exception)" cases in ICU4J (more than 50, while there are less than 90 "catch (MissingResourceError)" cases, for a little bit of an apples&oranges contrast. ("Fruitbowl contrast"? :-P)